js/background.js

```diff
@@ -359,26 +365,43 @@ function updateHeaders(headers, settings) {
     return headers;
 }
 
-// add listeners for web requests:
-// - onBeforeSendHeaders
-// - onHeadersReceived
-// so they can be edited on-the-fly to enable API calls from plug-ins
-// https://developer.chrome.com/docs/extensions/reference/webRequest/
+/** 
+* add listeners for web requests:
+* - onBeforeSendHeaders
+* - onHeadersReceived
+* so they can be edited on-the-fly to enable API calls from plug-ins
+* https://developer.chrome.com/docs/extensions/reference/webRequest/
+* https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest
+*/
+
 function addWebRequestListeners() {
-    if (!chrome.webRequest.onBeforeSendHeaders.hasListeners())
-        chrome.webRequest.onBeforeSendHeaders.addListener(updateRequestHeaders, { urls : ["<all_urls>"] }, ["blocking", "requestHeaders", "extraHeaders"]);
-    if (!chrome.webRequest.onHeadersReceived.hasListeners())
-        chrome.webRequest.onHeadersReceived.addListener(updateResponseHeaders, { urls : ["<all_urls>"] }, ["blocking", "responseHeaders", "extraHeaders"]);
+    if (!chrome.webRequest.onBeforeSendHeaders.hasListener(updateRequestHeaders)) {
+        chrome.webRequest.onBeforeSendHeaders.addListener(updateRequestHeaders, { urls : ["<all_urls>"] }, [
+            'blocking',
+            'requestHeaders',
+            chrome.webRequest.OnSendHeadersOptions.EXTRA_HEADERS,
+        ].filter(Boolean));
+    }
+    if (!chrome.webRequest.onHeadersReceived.hasListener(updateResponseHeaders)){
+        chrome.webRequest.onHeadersReceived.addListener(updateResponseHeaders, { urls : ["<all_urls>"] }, [
+            'blocking',
+            'responseHeaders',
+            chrome.webRequest.OnSendHeadersOptions.EXTRA_HEADERS,
+        ].filter(Boolean));
+    }
+    
 }
 
-// remove listeners for web requests:
-// - onBeforeSendHeaders
-// - onHeadersReceived
-// also remove headers settings since they are not used anymore
+/**
+* remove listeners for web requests:
+* - onBeforeSendHeaders
+* - onHeadersReceived
+* also remove headers settings since they are not used anymore
+*/
 function removeWebRequestListeners() {
-    if (chrome.webRequest.onBeforeSendHeaders.hasListeners())
+    if (!chrome.webRequest.onBeforeSendHeaders.hasListener(updateRequestHeaders))
         chrome.webRequest.onBeforeSendHeaders.removeListener(updateRequestHeaders);
-    if (chrome.webRequest.onHeadersReceived.hasListeners())
+    if (!chrome.webRequest.onHeadersReceived.hasListener(updateResponseHeaders))
         chrome.webRequest.onHeadersReceived.removeListener(updateResponseHeaders);
```

Comment:
This doesn't look right: if there is no listener then you are removing them? Probably a copy/paste typo.

Reply:
oh, yeah. It is